tamr_toolbox/utils/downstream.py

```diff
@@ -0,0 +1,245 @@
+            either based on backened pipeline (if project still exists) or based on
+            regex (dataset name has suffix 'unified_dataset').
+
+    Returns:
+        List of Dataset objects ordered by number of its downstream dependencies.
+            Note that there can be bidirectional dependency so datasets with same number of
+            dependencies can depend on each other.
+    """
+    downstream_datasets = _find_downstream_datasets(
+        dataset.name,
+        include_dependencies_by_name=include_dependencies_by_name,
+        client=dataset.client,
+    )
+    # Remove the target dataset from the list
+    downstream_datasets.pop(dataset.name)
+    # Extract the list of the Dataset objects
+    downstream_datasets = list(downstream_datasets.values())
+    # sort list of downstream datasets by number of dependencies
+    # because parent dataset always has greater number of dependencies than its child datasets
+    # note that there can be bidirectional dependency so datasets with same number of
+    # dependencies can depend on each other
+    LOGGER.info("Sorting the downstream datasets by number of dependencies")
+    sorted_datasets = sorted(downstream_datasets, key=lambda x: _count_dependencies(x))
+    return sorted_datasets
+
+
+def projects(dataset: Dataset, *, include_dependencies_by_name: bool = False) -> List[Project]:
+    """Return list of downstream project_list for a dataset.
+
+    Args:
+        dataset: The target dataset.
+        include_dependencies_by_name: A Boolean indicating whether to return datasets with
+            dependencies suggested by name matching. Default to False.
+            No dependencies will be found by name if the `dataset` is not an unified dataset
+            either based on backened pipeline (if project still exists) or based on
+            regex (dataset name has suffix 'unified_dataset').
+
+    Returns:
+        List of downstream project_list in order,
+            including the project the target `dataset` is part of.
+
+    """
+    downstream_datasets = datasets(
+        dataset, include_dependencies_by_name=include_dependencies_by_name
+    )
+    project_ids = []
+    # downstream dataset list is ordered by number of dependency ascending
+    for dependency in downstream_datasets:
+        temp = _find_associated_projects(dependency)
+        project_ids.extend(temp)
+    # add project_list of the target dataset
+    project_ids.extend(_find_associated_projects(dataset))
+    LOGGER.debug(project_ids)
+    # use Ordered Dict to keep the order (set doesn't keep the order)
+    project_ids = OrderedDict().fromkeys(project_ids)
+    # Need to reverse to get the most upstream project first
+    project_list = [
+        dataset.client.projects.by_resource_id(i) for i in reversed(project_ids.keys())
+    ]
+    return project_list
+
+
+def _find_downstream_datasets(
+    dataset_name,
```

Comment:
You're missing a typehint for dataset_name. Also within the docstring you reference it as  "dataset" not "dataset_name"